pkg/api/authz.go

```diff
@@ -0,0 +1,222 @@
+package api
+
+import (
+	"context"
+	"encoding/base64"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/anuvu/zot/pkg/log"
+	"github.com/gorilla/mux"
+)
+
+type key int
+
+const (
+	CREATE           = "create"
+	READ             = "read"
+	UPDATE           = "update"
+	DELETE           = "delete"
+	contextKeyID key = 0
+)
+
+type AccessControlConfig struct {
+	Repositories Repositories
+	AdminPolicy  Policy
+}
+
+type Repositories map[string]PolicyGroup
+
+type Repository struct {
+	Name          string
+	Policies      []Policy
+	DefaultPolicy []string
+}
+
+type PolicyGroup struct {
+	Policies      []Policy
+	DefaultPolicy []string
+}
+
+type Policy struct {
+	Users   []string
+	Actions []string
+}
+
+// AccessController authorizes users to act on resources.
+type AccessController struct {
+	Config *AccessControlConfig
+	Log    log.Logger
+}
+
+// AccessControllerContext context passed down to http.Handlers.
+type AccessControllerContext struct {
+	userAllowedRepos []string
+	isAdmin          bool
+}
+
+func NewAccessController(config *Config) *AccessController {
+	return &AccessController{
+		Config: config.AccessControl,
+		Log:    log.NewLogger(config.Log.Level, config.Log.Output),
+	}
+}
+
+// getReadRepos get repositories from config file that the user has READ perms.
+func (a *AccessController) getReadRepos(username string) []string {
+	var repos []string
+
+	for r, pg := range a.Config.Repositories {
+		for _, p := range pg.Policies {
+			if (contains(p.Users, username) && contains(p.Actions, READ)) ||
+				contains(pg.DefaultPolicy, READ) {
+				repos = append(repos, r)
+			}
+		}
+	}
+
+	return repos
+}
+
+// can verify if a user can do action on repository.
+func (a *AccessController) can(username, action, repository string) bool {
+	can := false
+	// check repo based policy
+	for r, pg := range a.Config.Repositories {
+		if strings.Contains(repository, r) {
```

Comment:
If a particular user have write permission for this repo:- /infra/ubuntu and he tries to request to write ubuntu (a new repo e.g ```skopeo copy docker://ubuntu:latest docker://<zot-address>/ubuntu:latest). I think user will be able to write on our root directory since we are using strings.Contains(repository, r) ?

Reply:
That's right, thank you, I'll fix it.